include/nbl/ui/IInputEventChannel.h

```diff
@@ -0,0 +1,108 @@
+#ifndef __NBL_I_INPUT_EVENT_CHANNEL_H_INCLUDED__
+#define __NBL_I_INPUT_EVENT_CHANNEL_H_INCLUDED__
+
+#include <mutex>
+
+#include "nbl/core/IReferenceCounted.h"
+#include "nbl/core/containers/CCircularBuffer.h"
+#include "nbl/core/SRange.h"
+
+namespace nbl {
+namespace ui
+{
+
+class IInputEventChannel : public core::IReferenceCounted
+{
+protected:
+    virtual ~IInputEventChannel() = default;
+
+public:
+    // TODO Any other common (not depending on event struct type) member functions?
+
+    virtual bool empty() const = 0;
+};
+
+namespace impl
+{
+    template <typename EventType>
+    class IEventChannelBase : public IInputEventChannel
+    {
+    protected:
+        static inline constexpr size_t MaxEvents = 1024u;
+
+        using cb_t = core::CCompileTimeSizedCircularBuffer<EventType, MaxEvents>;
+        using iterator_t = typename cb_t::iterator;
+        using range_t = core::SRange<EventType, iterator_t, iterator_t>;
+
+        std::mutex m_bgEventBufMtx;
+        cb_t m_bgEventBuf;
+        cb_t m_frontEventBuf;
+
+        // Lock while working with background event buffer
+        std::unique_lock<std::mutex> lockBackgroundBuffer()
+        {
+            return std::unique_lock<std::mutex>(m_bgEventBufMtx);
+        }
+
+        // Use this within OS-specific impl (Windows callback/XNextEvent loop thread/etc...)
+        void pushIntoBackground(EventType&& ev)
+        {
+            m_bgEventBuf.push_back(std::move(ev));
+        }
+
+        // WARNING: Access to getEvents() must be externally synchronized to be safe!
+        // (the function returns range of iterators)
+        range_t getEvents()
+        {
+            downloadFromBackgroundIntoFront();
+            return range_t(m_frontEventBuf.begin(), m_frontEventBuf.end());
+        }
+
+        virtual ~IEventChannelBase() = default;
+
+    private:
+        void downloadFromBackgroundIntoFront()
+        {
+            auto lk = lockBackgroundBuffer();
```

Comment:
call it a transfer or download lock, cause its sole usage is to ensure that multiple threads dont try to push to the front buffer.

there's no locking going on for the background buffer really (multiple threads are pushing events, single thread is reading - and this is the only thing thats getting locked). Its entirely possible and well behaved to have pushes going on at the same time pops are occuring.

P.S. You need to make push/pop "atomic" for the overflowable buffer, dont want a situation where move assignment during pop is interrupted by a push onto the same slot (this would be really bad for non PoD types which actually have ctors and dtors, for obvious reasons). Naturally not a concern for the blocking buffer because it cannot happen.

Reply:
hmm you could even call it a foreground buffer lock and give wider access to it, because if you have a transfer lock, you'll also need a foreground buffer lock to ensure others can externally serialize the read access with the write access of the `downloadFromBackgroundIntoFront` method.

So instead of ending up with 2 locks (transfer and foreground), lets just have a foreground lock.